lib/recursive-methods.rb

```diff
@@ -1,45 +1,86 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+
+    raise ArgumentError if n < 0
+
+    if n == 0 #base case
+        return 1
+    else
+        return n * factorial(n-1)
+    end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n^2)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    if s.length <= 1
+        return s
+    else
+        return s[-1] + reverse(s[1..-2]) + s[0]
+    end
 end
 
+
+
 # Time complexity: ?
 # Space complexity: ?
 def reverse_inplace(s)
```

Comment:
⚠️ This is not in place.  You are creating new strings with each recursive call.  Instead of modifying the original string.

Think about adjusting the method signature to this.  

```ruby
def reverse_inplace(s, low = 0, high = s.length - 1)
```

Does this help you see a way to do it in-place.